snuba/datasets/processors/spans_processor.py

```diff
@@ -0,0 +1,484 @@
+        processed_span["transaction_id"] = None
+        processed_span["trace_id"] = str(uuid.UUID(span_dict["trace_id"]))
+        processed_span["span_id"] = int(span_dict["span_id"], 16)
+        processed_span["parent_span_id"] = int(span_dict.get("parent_span_id", 0), 16)
+        processed_span["is_segment"] = 0
+        processed_span["op"] = _unicodify(span_dict.get("op", ""))
+        processed_span["group"] = int(span_dict.get("hash", 0), 16)
+        processed_span["exclusive_time"] = span_dict.get("exclusive_time", 0)
+        processed_span["description"] = _unicodify(span_dict.get("description", ""))
+
+        start_timestamp = span_dict["start_timestamp"]
+        end_timestamp = span_dict["timestamp"]
+        processed_span["start_timestamp"], _ = self.__extract_timestamp(start_timestamp)
+        if end_timestamp - start_timestamp < 0:
+            # Seems we have some negative durations in the DB
+            metrics.increment("negative_duration")
+
+        processed_span["end_timestamp"], _ = self.__extract_timestamp(end_timestamp)
+        duration_secs = (
+            processed_span["end_timestamp"] - processed_span["start_timestamp"]
+        ).total_seconds()
+        processed_span["duration"] = max(int(duration_secs * 1000), 0)
+        processed_span["exclusive_time"] = span_dict.get("exclusive_time", 0)
+        tags: Mapping[str, Any] = _as_dict_safe(span_dict.get("data", None))
+        span_tag_keys, span_tag_values = extract_extra_tags(tags)
+        processed_span["tags.key"].extend(span_tag_keys)
+        processed_span["tags.value"].extend(span_tag_values)
+
+        self._process_child_span_module_details(processed_span, span_dict)
+
+        # The processing is of modules does not guarantee that all clickhouse columns would be
+        # setup. Just to be safe, lets setup all required clickhouse columns (which are not
+        # nullable) with their default values if they have not been setup yet.
+        processed_span["span_kind"] = processed_span.get("span_kind", "")
+        processed_span["span_status"] = processed_span.get("span_status", 0)
+        processed_span["measurements.key"] = processed_span.get("measurements.key", [])
+        processed_span["measurements.value"] = processed_span.get(
+            "measurements.value", []
+        )
+        processed_span["module"] = processed_span.get("module", "")
+
+        return processed_span
+
+    def _process_spans(
+        self, event_dict: EventDict, common_span_fields: CommonSpanDict
+    ) -> MutableSequence[MutableMapping[str, Any]]:
+        data = event_dict["data"]
+        processed_spans: MutableSequence[MutableMapping[str, Any]] = []
+        for span in data.get("spans", []):
+            processed_span = self._process_span(span, common_span_fields)
+            if processed_span is not None:
+                processed_spans.append(processed_span)
+
+        return processed_spans
+
+    def process_message(
```

Comment:
Since this is an experiment and the incoming data will probably change format a few times, I might wrap more of this in Exception handling code so the consumer doesn't die every time we get a bad message.

Reply:
That's a good point. You are correct that this has potential to blow up. We do not want to care about this blowing up just yet.

Reply:
The processing code has been wrapped in a try except block [here](https://github.com/getsentry/snuba/pull/4102/files#diff-c76e01559784d4d35d2a011e4430e38f324f617fc24180607d8aec538241238cR470-R491)